server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -607,7 +607,15 @@ private static void parseDynamicValue(final DocumentParserContext context, Strin
             failIfMatchesRoutingPath(context, currentFieldName);
             return;
         }
-        context.dynamic().getDynamicFieldsBuilder().createDynamicFieldFromValue(context, currentFieldName);
+        try {
+            context.dynamic().getDynamicFieldsBuilder().createDynamicFieldFromValue(context, currentFieldName);
+        } catch (IllegalArgumentException e) {
```

Comment:
is it guaranteed that this exception is caused by reaching the total fields limit? This seems very delicate to me.

Reply:
I was thinking of adding a more specific exception class for this but that would change the response we're sending in case of errors from `type: illegal_argument_exception` to something else, such as `field_limit_exceeded_exception`. Not sure if that's a potentially breaking change.

Reply:
I am not convinced that catching the exception is the way to introduce this new behaviour. I would rather look at supporting it deeper when we try to dynamically map fields. I would need to spend more time looking at this myself to come up with a more concrete proposal.

Reply:
Makes sense. I gave it a shot in 0634d269531d9736f47aff4e3b8752f30a665078